From dfdfe39e5b5357af8336cf28abe0b5df3897c767 Mon Sep 17 00:00:00 2001
From: Josue David Hernandez Gutierrez <josue.d.hernandez.gutierrez@intel.com>
Date: Fri, 26 Oct 2018 16:46:37 -0500
Subject: [PATCH 35/42] add stateless for locale files locale.conf
 00-keyboard.conf

systemd must take this files from /etc and in case there are not files
there /usr/share/defaults/etc

Signed-off-by: Josue David Hernandez Gutierrez <josue.d.hernandez.gutierrez@intel.com>
---
 src/locale/keymap-util.c | 42 ++++++++++++++++++++++++++++++----------
 1 file changed, 32 insertions(+), 10 deletions(-)

diff --git a/src/locale/keymap-util.c b/src/locale/keymap-util.c
index f8c36c94f5..55a3e7ee44 100644
--- a/src/locale/keymap-util.c
+++ b/src/locale/keymap-util.c
@@ -93,6 +93,7 @@ void locale_simplify(char *locale[_VARIABLE_LC_MAX]) {
 int locale_read_data(Context *c, sd_bus_message *m) {
         struct stat st;
         int r;
+        bool etc_conf = false;
 
         /* Do not try to re-read the file within single bus operation. */
         if (m) {
@@ -104,6 +105,11 @@ int locale_read_data(Context *c, sd_bus_message *m) {
         }
 
         r = stat("/etc/locale.conf", &st);
+        if (r >= 0)
+                etc_conf = true;
+        else if (errno == ENOENT)
+                r = stat("/usr/share/defaults/etc/locale.conf", &st);
+
         if (r < 0 && errno != ENOENT)
                 return -errno;
 
@@ -118,7 +124,9 @@ int locale_read_data(Context *c, sd_bus_message *m) {
                 c->locale_mtime = t;
                 context_free_locale(c);
 
-                r = parse_env_file(NULL, "/etc/locale.conf",
+                r = parse_env_file(NULL,
+                                   (etc_conf)?"/etc/locale.conf":
+                                              "/usr/share/defaults/etc/locale.conf",
                                    "LANG",              &c->locale[VARIABLE_LANG],
                                    "LANGUAGE",          &c->locale[VARIABLE_LANGUAGE],
                                    "LC_CTYPE",          &c->locale[VARIABLE_LC_CTYPE],
@@ -162,6 +170,7 @@ int vconsole_read_data(Context *c, sd_bus_message *m) {
         struct stat st;
         usec_t t;
         int r;
+        bool etc_conf = false;
 
         /* Do not try to re-read the file within single bus operation. */
         if (m) {
@@ -172,14 +181,19 @@ int vconsole_read_data(Context *c, sd_bus_message *m) {
                 c->vc_cache = sd_bus_message_ref(m);
         }
 
-        if (stat("/etc/vconsole.conf", &st) < 0) {
-                if (errno != ENOENT)
-                        return -errno;
+        r = stat("/etc/vconsole.conf", &st);
+        if (r >= 0) {
+                etc_conf = true;
+        } else if (errno == ENOENT)
+                r = stat("/usr/share/defaults/etc/vconsole.conf", &st);
 
+        if (r < 0 && errno == ENOENT) {
                 c->vc_mtime = USEC_INFINITY;
                 context_free_vconsole(c);
                 return 0;
         }
+        else if (r < 0)
+                return -errno;
 
         /* If mtime is not changed, then we do not need to re-read */
         t = timespec_load(&st.st_mtim);
@@ -189,7 +203,9 @@ int vconsole_read_data(Context *c, sd_bus_message *m) {
         c->vc_mtime = t;
         context_free_vconsole(c);
 
-        r = parse_env_file(NULL, "/etc/vconsole.conf",
+        r = parse_env_file(NULL, (etc_conf)?
+                           "/etc/vconsole.conf":
+                               "/usr/share/defaults/etc/vconsole.conf",
                            "KEYMAP",        &c->vc_keymap,
                            "KEYMAP_TOGGLE", &c->vc_keymap_toggle);
         if (r < 0)
@@ -204,6 +220,7 @@ int x11_read_data(Context *c, sd_bus_message *m) {
         struct stat st;
         usec_t t;
         int r;
+        bool etc_conf = false;
 
         /* Do not try to re-read the file within single bus operation. */
         if (m) {
@@ -214,14 +231,18 @@ int x11_read_data(Context *c, sd_bus_message *m) {
                 c->x11_cache = sd_bus_message_ref(m);
         }
 
-        if (stat("/etc/X11/xorg.conf.d/00-keyboard.conf", &st) < 0) {
-                if (errno != ENOENT)
-                        return -errno;
+        r = stat("/etc/X11/xorg.conf.d/00-keyboard.conf", &st);
+        if (r >= 0) {
+                etc_conf = true;
+        } else if (errno == ENOENT)
+                r = stat("/usr/share/defaults/etc/X11/xorg.conf.d/00-keyboard.conf", &st);
 
+        if (r < 0 && errno == ENOENT) {
                 c->x11_mtime = USEC_INFINITY;
                 context_free_x11(c);
                 return 0;
-        }
+        } else if (r < 0)
+                return -errno;
 
         /* If mtime is not changed, then we do not need to re-read */
         t = timespec_load(&st.st_mtim);
@@ -231,7 +252,8 @@ int x11_read_data(Context *c, sd_bus_message *m) {
         c->x11_mtime = t;
         context_free_x11(c);
 
-        f = fopen("/etc/X11/xorg.conf.d/00-keyboard.conf", "re");
+        f = fopen((etc_conf)?"/etc/X11/xorg.conf.d/00-keyboard.conf":
+                             "/usr/share/defaults/etc/X11/xorg.conf.d/00-keyboard.conf", "re");
         if (!f)
                 return -errno;
 
-- 
2.23.0

